pxtcompiler/emitter/backthumb.ts

```diff
@@ -123,6 +143,73 @@ ${lbl}:`
 `
         }
 
+        arithmetic() {
+            let r = ""
+
+            if (!target.taggedInts || target.boxDebug) {
+                return r
+            }
+
+            for (let op of ["adds", "subs", "ands", "orrs", "eors"]) {
+                r +=
+                    `
+_numops_${op}:
+    @scope _numops_${op}
+    lsls r2, r0, #31
+    beq .boxed
+    lsls r2, r1, #31
+    beq .boxed
+`
+                if (op == "adds" || op == "subs")
+                    r += `
+    subs r2, r1, #1
+    ${op} r2, r0, r2
+    bvs .boxed
+    movs r0, r2
+    blx lr
+`
+                else {
+                    r += `    ${op} r0, r1\n`
+                    if (op == "eors")
+                        r += `    adds r0, r0, #1\n`
+                    r += `    blx lr\n`
+                }
+
+                r += `
+.boxed:
+    push {lr}
+    bl numops::${op}
+    pop {pc}
+`
+            }
+
+            r += `
+@scope _numops_toInt
+_numops_toInt:
+    asrs r0, r0, #1
+    bcc .over
+    blx lr
+.over:
+    push {lr}
+    bl pxt::toInt
+    pop {pc}
+
+_numops_fromInt:
+    lsls r2, r0, #1
+    asrs r1, r2, #1
+    cmp r0, r1
+    bne .over2
+    adds r0, r2, #1
```

Comment:
4 instructions instead of 5?

teq r0, r0, LSL#1
bmi .over2
lsl r0, r0, #1
add r0, r0, #1